File: app/src/main/java/com/giaquino/android/cropped/ui/topics/TopicsAdapter.java
package com.giaquino.android.cropped.ui.topics;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.giaquino.android.cropped.databinding.TopicsItemBinding;
import com.giaquino.android.cropped.data.model.Topic;

import java.util.ArrayList;
import java.util.List;

public class TopicsAdapter extends RecyclerView.Adapter<TopicsAdapter.TopicsViewHolder> {

    private List<Topic> topics = new ArrayList<>();
    private TopicsViewHolder.OnItemClickListener itemClickListener;

    public TopicsAdapter(TopicsViewHolder.OnItemClickListener itemClickListener) {
        this.itemClickListener = itemClickListener;
    }

    @NonNull
    @Override
    public TopicsViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return TopicsViewHolder.create(parent);
    }

    @Override
    public void onBindViewHolder(@NonNull TopicsViewHolder holder, int position) {
        holder.bind(topics.get(position), itemClickListener);
    }

    @Override
    public int getItemCount() {
        return topics.size();
    }

    public void setTopics(List<Topic> topics) {
        this.topics.clear();
        this.topics.addAll(topics);
        this.notifyDataSetChanged();
    }

    public static class TopicsViewHolder extends RecyclerView.ViewHolder {

        public interface OnItemClickListener {
            void onItemClick(Topic topic);
        }

        private TopicsItemBinding binding;

        public static TopicsViewHolder create(ViewGroup parent) {
            TopicsItemBinding binding = TopicsItemBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
            return new TopicsViewHolder(binding);
        }

        public TopicsViewHolder(TopicsItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(Topic topic, OnItemClickListener listener) {
            binding.title.setText(topic.title);
            binding.subtitle.setText("By " + topic.owners.get(0).name);
            binding.description.setText(topic.description);
            binding.contributions.setText(getDisplayableTotalPhotos(topic.totalPhotos) + " contributions");

            Glide.with(binding.banner).load(topic.coverPhoto.urls.regular).into(binding.banner);
            Glide.with(binding.avatar).load(topic.owners.get(0).profileImage.large).circleCrop().into(binding.avatar);

            binding.card.setOnClickListener(v -> listener.onItemClick(topic));
        }

        private String getDisplayableTotalPhotos(int totalPhotos) {
            String total = "";
            if (totalPhotos >= 1000) {
                int temp = totalPhotos / 1000;
                total += temp;
                temp = (totalPhotos - (temp * 1000)) / 100;
                if (temp > 0) {
                    total += "." + temp + "k";
                }
            } else {
                return String.valueOf(totalPhotos);
            }
            return total;
        }
    }
}

File: app/src/main/java/com/giaquino/android/cropped/ui/topics/TopicsAdapter.kt
package com.giaquino.android.cropped.ui.topics

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.giaquino.android.cropped.data.model.Topic
import com.giaquino.android.cropped.databinding.TopicsItemBinding
import com.giaquino.android.cropped.ui.topics.TopicsAdapter.TopicsViewHolder
import java.util.*

class TopicsAdapter(private val listener : (Topic) -> Unit) : RecyclerView.Adapter<TopicsViewHolder>() {

    private val topics: MutableList<Topic> = ArrayList()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TopicsViewHolder {
        return TopicsViewHolder.create(parent)
    }

    override fun onBindViewHolder(holder: TopicsViewHolder, position: Int) {
        holder.bind(topics[position], listener)
    }

    override fun getItemCount(): Int {
        return topics.size
    }

    fun setTopics(topics: List<Topic>) {
        this.topics.clear()
        this.topics.addAll(topics)
        notifyDataSetChanged()
    }

    class TopicsViewHolder(private val binding: TopicsItemBinding) : RecyclerView.ViewHolder(binding.root) {
        interface OnItemClickListener {
            fun onItemClick(topic: Topic?)
        }

        fun bind(topic: Topic, listener : (Topic) -> Unit) {
            binding.apply {
                title.text = topic.title
                subtitle.text = "By ${topic.owners[0].name}"
                description.text = topic.description
                contributions.text = "${getDisplayableTotalPhotos(topic.totalPhotos)} contributions"
                card.setOnClickListener { listener(topic) }
            }
            Glide.with(binding.banner).load(topic.coverPhoto.urls.regular).into(binding.banner)
            Glide.with(binding.avatar).load(topic.owners[0].profileImage.large).circleCrop().into(binding.avatar)
        }

        private fun getDisplayableTotalPhotos(totalPhotos: Int): String {
            var total = ""
            if (totalPhotos >= 1000) {
                var temp = totalPhotos / 1000
                total += temp
                temp = (totalPhotos - temp * 1000) / 100
                if (temp > 0) {
                    total += "." + temp + "k"
                }
            } else {
                return totalPhotos.toString()
            }
            return total
        }

        companion object {
            fun create(parent: ViewGroup): TopicsViewHolder {
                return TopicsViewHolder(TopicsItemBinding.inflate(LayoutInflater.from(parent.context), parent, false))
            }
        }
    }
}